consensus/consensus_v2.go

```diff
@@ -607,60 +628,64 @@ func (consensus *Consensus) tryCatchup() {
 
 		preparedMsgs := consensus.pbftLog.GetMessagesByTypeSeqHash(msg_pb.MessageType_PREPARED, msgs[0].BlockNum, msgs[0].BlockHash)
 		if len(preparedMsgs) > 1 {
-			utils.GetLogInstance().Info("[PBFT] we should only get one prepared message for a given blockNum", "blockNum", consensus.blockNum, "numMsgs", len(preparedMsgs))
+			utils.GetLogInstance().Warn("[PBFT] we get more than one prepared messages for a given blockNum", "blockNum", consensus.blockNum, "numMsgs", len(preparedMsgs))
 		}
 		if len(preparedMsgs) == 0 {
 			break
 		}
 
-		cnt := 0
-		for _, msg := range preparedMsgs {
-			if msg.BlockHash == msgs[0].BlockHash {
-				cnt++
-				consensus.blockHash = [32]byte{}
-				consensus.blockNum = consensus.blockNum + 1
-				consensus.viewID = consensus.viewID + 1
-
-				//#### Read payload data from committed msg
-				aggSig := make([]byte, 48)
-				bitmap := make([]byte, len(msgs[0].Payload)-48)
-				offset := 0
-				copy(aggSig[:], msgs[0].Payload[offset:offset+48])
-				offset += 48
-				copy(bitmap[:], msgs[0].Payload[offset:])
-				//#### END Read payload data from committed msg
-
-				//#### Read payload data from prepared msg
-				prepareSig := make([]byte, 48)
-				prepareBitmap := make([]byte, len(msg.Payload)-48)
-				offset = 0
-				copy(prepareSig[:], msg.Payload[offset:offset+48])
-				offset += 48
-				copy(prepareBitmap[:], msg.Payload[offset:])
-				//#### END Read payload data from committed msg
-
-				// Put the signatures into the block
-				block.SetPrepareSig(prepareSig, prepareBitmap)
-
-				block.SetCommitSig(aggSig, bitmap)
-				utils.GetLogInstance().Info("Adding block to chain", "numTx", len(block.Transactions()))
-				consensus.OnConsensusDone(block)
-				consensus.ResetState()
-
-				select {
-				case consensus.VerifiedNewBlock <- block:
-				default:
-					utils.GetLogInstance().Info("[SYNC] consensus verified block send to chan failed", "blockHash", block.Hash())
-					continue
-				}
-
-				break
-			}
-		}
-		if cnt == 0 { // didn't find match block for current blockNum, return
+		msg := consensus.pbftLog.FindMessageByMaxViewID(preparedMsgs)
+		if msg == nil {
 			break
 		}
+		consensus.blockHash = [32]byte{}
```

Comment:
Why these are moved out of loop?

Reply:
In previous code, the loop is unnecessary, especially I add FindMessageByMaxViewID. i.e. we only choose the latest viewID for prepared message.